cpp/tests/traversal/tsp_test.cu

```diff
@@ -0,0 +1,205 @@
+ public:
+  Tests_Tsp() {}
+  static void SetupTestCase() {}
+  static void TearDownTestCase() {}
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  void run_current_test(const Tsp_Usecase& param)
+  {
+    const ::testing::TestInfo* const test_info =
+      ::testing::UnitTest::GetInstance()->current_test_info();
+    std::stringstream ss;
+    std::string test_id = std::string(test_info->test_case_name()) + std::string(".") +
+                          std::string(test_info->name()) + std::string("_") +
+                          cugraph::test::getFileName(param.tsp_file) + std::string("_") +
+                          ss.str().c_str();
+
+    float tol = 10E-2f;
+    HighResClock hr_clock;
+    double time_tmp;
+    Route input;
+
+    std::cout << "File: " << param.tsp_file.c_str() << "\n";
+    int nodes = load_tsp(param.tsp_file.c_str(), &input);
+
+    // Device alloc
+    raft::handle_t handle;
+    rmm::device_uvector<int> vertices(static_cast<size_t>(nodes), nullptr);
+    rmm::device_uvector<int> route(static_cast<size_t>(nodes), nullptr);
+    rmm::device_uvector<float> x_pos(static_cast<size_t>(nodes), nullptr);
+    rmm::device_uvector<float> y_pos(static_cast<size_t>(nodes), nullptr);
+
+    int* vtx_ptr   = vertices.data();
+    int* d_route   = route.data();
+    float* d_x_pos = x_pos.data();
+    float* d_y_pos = y_pos.data();
+
+    CUDA_TRY(cudaMemcpy(vtx_ptr, input.cities.data(), sizeof(int) * nodes, cudaMemcpyHostToDevice));
+    CUDA_TRY(
+      cudaMemcpy(d_x_pos, input.x_pos.data(), sizeof(float) * nodes, cudaMemcpyHostToDevice));
+    CUDA_TRY(
+      cudaMemcpy(d_y_pos, input.y_pos.data(), sizeof(float) * nodes, cudaMemcpyHostToDevice));
+
+    // Default parameters
+    int restarts     = 4096;
+    bool beam_search = true;
+    int k            = 4;
+    int nstart       = 0;
+    bool verbose     = false;
+
+    hr_clock.start();
+    cudaDeviceSynchronize();
+    cudaProfilerStart();
+
+    float final_cost = cugraph::traveling_salesman(
+      handle, vtx_ptr, d_route, d_x_pos, d_y_pos, nodes, restarts, beam_search, k, nstart, verbose);
+    cudaProfilerStop();
+    cudaDeviceSynchronize();
+    hr_clock.stop(&time_tmp);
+    std::cout << "tsp_time: " << time_tmp << " us" << std::endl;
+    std::cout << "Ref cost is: " << param.ref_cost << "\n";
+    std::cout << "Final cost is: " << final_cost << "\n";
+    float err = fabs(final_cost - param.ref_cost);
+    err /= param.ref_cost;
+    std::cout << "Approximation error is: " << err * 100 << "%\n";
+    EXPECT_LE(err, tol);
+  }
+
+ private:
+  std::vector<std::string> split(const std::string& s, char delimiter)
+  {
+    std::vector<std::string> tokens;
+    std::string token;
+    std::istringstream tokenStream(s);
+    while (std::getline(tokenStream, token, delimiter)) {
+      if (token.size() == 0) continue;
+      tokens.push_back(token);
+    }
+    return tokens;
+  }
+
+  int load_tsp(const char* fname, Route* input)
```

Comment:
Why not use mtx or csv files?

Reply:
We want to compare against [TSPLIB](http://comopt.ifi.uni-heidelberg.de/software/TSPLIB95/) which solves TSP as a city problem, so the format corresponds to 2D positions as input.

Reply:
I see how it would be relevant in the context of a demo/benchmark. Not sure if that's a strong enough motivation in the context cugraph testing though. It seems to me it is going in the opposite direction than what @rlratzel is trying to do for making testing more generic.

Reply:
Added the fixme in this commit: https://github.com/rapidsai/cugraph/pull/1360/commits/f85cb0d6d5a8f7bafef61891af3d03ec1f3d43c2